torch/utils/checkpoint.py

```diff
@@ -335,7 +340,370 @@ def forward(input):
+#   4. Continue original forward at (4), not saving x and w.
+#   5. (5) triggers a recompute of fn
+#   6. During recompute, we see that in the original graph, gx has already
+#      cleared x and y since backward is run at (3) without retain_graph=True
+#      We save x and w, however.
+#   7. Continue with returning
+#
+# [ backward within nested checkpoint ]
+#
+# Another case to consider is when we do backward within checkpoint, but we are
+# also in a nested checkpoint. Properly handling this case has implications for
+# how inputs are saved.
+#
+# def f(x):
+#   y = x.sin()
+#   z = y.cos()
+#   gx, = torch.autograd.grad(z, x)       # (1)
+#   return z
+#
+# def g(x):
+#   return checkpoint(f)(x)
+#
+# out = checkpoint(g)(inp)
+#
+# In the above example, when we recompute for the outer checkpoint (the one
+# wrapping g), we are recomputing checkpointed f.
+#
+# When checkpointed f was original computed in forward, there was already a
+# checkpoint active when we entered f's checkpoint, which means that f's
+# checkpoint is nested. However, during the recomputation of g, we enter
+# into a checkpoint wrapping f yet again, but this time there is no longer a
+# checkpoint active. How should we save f's inputs in this situation?
+#
+# Recall that we should save f's inputs onto the parent checkpoint if we are
+# nested, or directly onto the checkpoint otherwise, so on the surface it sounds
+# like we should save directly onto the checkpoint.
+#
+# Strangely, the answer here is actually both!
+#
+# In addition to saving our inputs onto the frame directly, we also save f's
+# inputs onto the parent checkpoint. We do this for two reasons:
+#
+# 1) First, if we did not save f's inputs directly onto its own frame we would
+#    not be able to recompute f during recomputation for - see line marked (1)
+#    above in the example.
+# 2) Second, since f's checkpoint was originally saved its inputs as if it were
+#    nested, it must also save its inputs as if it were nested so that the
+#    indices of the recomputed variables match.
+#
+
+# NB: This is temporary and should be removed in a follow up PR. Early stopping
+#     is currently disabled by default. Since some nested test cases require
+#     ealry stopping to pass, _set_checkpoint_early_stop can be used to enable.
+_enable_checkpoint_early_stop = False
+
+@contextlib.contextmanager
+def _set_checkpoint_early_stop(enable):
+    global _enable_checkpoint_early_stop
+    try:
+        prev = _enable_checkpoint_early_stop
+        _enable_checkpoint_early_stop = enable
+        yield
+    finally:
+        _enable_checkpoint_early_stop = prev
+
+# See NOTE [ Nestable Checkpoint ] Rule #4
+class _Handle():
+    pass
+
+class _Holder():
+    def __init__(self, handle: _Handle):
+        self.handle: _Handle = handle
+
+# Reimplementation of torch.distributed.utils.{_pack,_unpack}_kwargs to avoid a import cycle
+def _pack_kwargs(*args: Any, **kwargs: Any) -> Tuple[Tuple[Any, ...], Tuple[str, ...]]:
```

Comment:
Ok this was a thing because we are using custom Functions to save the inputs, and those didn't support kwargs. 

lol I didn't come up with the names, but from the other perspective, you are could imagine flat args as a more "compact" representation, so it would be considered packing (I'd prefer to keep the naming in sync with the distributed version though if we end up keeping this)

Reply:
I have a sense that this will not be needed very soon, so no need to worry about it ;)

Reply:
Hmm, couldn't get rid of yet.